redis/cluster.py

```diff
@@ -0,0 +1,1884 @@
+                            )
+
+                            if len(disagreements) > 5:
+                                raise RedisClusterException(
+                                    'startup_nodes could not agree on a valid'
+                                    ' slots cache: {0}'.format(
+                                        ", ".join(disagreements))
+                                )
+
+        if not startup_nodes_reachable:
+            raise RedisClusterException(
+                "Redis Cluster cannot be connected. Please provide at least "
+                "one reachable node. "
+            )
+
+        # Create Redis connections to all nodes
+        self.create_redis_connections(list(tmp_nodes_cache.values()))
+
+        fully_covered = self.check_slots_coverage(tmp_slots)
+        if not fully_covered:
+            if self._require_full_coverage:
+                # Despite the requirement that the slots be covered, there
+                # isn't a full coverage
+                raise RedisClusterException(
+                    'All slots are not covered after query all startup_nodes.'
+                    ' {0} of {1} covered...'.format(
+                        len(self.slots_cache), REDIS_CLUSTER_HASH_SLOTS)
+                )
+            else:
+                # The user set require_full_coverage to False.
+                # In case of full coverage requirement in the cluster's Redis
+                # configurations, we will raise an exception. Otherwise, we may
+                # continue with partial coverage.
+                # see Redis Cluster configuration parameters in
+                # https://redis.io/topics/cluster-tutorial
+                if not self._skip_full_coverage_check and \
+                        self.cluster_require_full_coverage(tmp_nodes_cache):
+                    raise RedisClusterException(
+                        'Not all slots are covered but the cluster\'s '
+                        'configuration requires full coverage. Set '
+                        'cluster-require-full-coverage configuration to no on '
+                        'all of the cluster nodes if you wish the cluster to '
+                        'be able to serve without being fully covered.'
+                        ' {0} of {1} covered...'.format(
+                            len(self.slots_cache), REDIS_CLUSTER_HASH_SLOTS)
+                    )
+
+        # Set the tmp variables to the real variables
+        self.nodes_cache = tmp_nodes_cache
+        self.slots_cache = tmp_slots
+        # Populate the startup nodes with all discovered nodes
+        self.populate_startup_nodes(self.nodes_cache.values())
+
+    def close(self):
+        for node in self.nodes_cache.values():
+            if node.redis_connection:
+                node.redis_connection.close()
+
+    def reset(self):
+        if self.read_load_balancer is not None:
+            self.read_load_balancer.reset()
```

Comment:
It's probably cheaper to run a try/catch here if the read_load_balancer is None instead. Python's try/catch is cheaper. This also follows the python rule "ask for forgiveness, not permission".